### ISSUE TITLE
Pack file is not released on repository.close()

### ISSUE THREAD

**Issue by arysin** [eclipse-jgit/jgit#66]
### Version

6.7.0

### Operating System

Windows

### Bug description

I clone a repository (with try/resouce on Git object) and after it's closed the git directory can't be deleted on Windows

### Actual behavior

I get a message
"C:\Users\user\AppData\Local\Temp\jgitRepository14654396429264831000\.git\objects\pack\pack-065d80f75516484c8e0c0a3631b7cdf620f27328.pack: The process cannot access the file because it is being used by another process"
I tried to force close but it does not help and the message states the repo is already closed:
```
                git.close();
                git.getRepository().close();
                FileUtils.deleteDirectory(tmpJgitPath);
```
close() called when useCnt is already zero for Repository

### Expected behavior

The pack file should be released on repository.close() and .git directory should be deleted

### Relevant log output

```Shell
It looks like repository.objectDatabase.packed.packList.value.packs[0].fd (RandomAccessFile) is still open.

From what I can see this is still the case in master: the org.eclipse.jgit.internal.storage.file.Pack.close() does not release the "fd".
```

**Comment by arysin** [comment]
The (ugly) workaround like this removes the error message:
```
                ObjectDatabase objectDatabase = git.getRepository().getObjectDatabase();
                Object packed = FieldUtils.getDeclaredField(ObjectDirectory.class, "packed", true).get(objectDatabase);
                AtomicReference packList = (AtomicReference) FieldUtils.getDeclaredField(packed.getClass(), "packList", true).get(packed);
                Object packList_ = packList.get();
                Pack[] packs = (Pack[]) FieldUtils.getDeclaredField(packList_.getClass(), "packs", true).get(packList_);
                for(Pack pack: packs) {
                    RandomAccessFile fd = (RandomAccessFile) FieldUtils.getDeclaredField(pack.getClass(), "fd", true).get(pack);
                    fd.close();
                }

```

**Comment by JoostK** [comment]
> From what I can see this is still the case in master: the org.eclipse.jgit.internal.storage.file.Pack.close() does not release the "fd".

There's some interaction with `WindowCache` that reaches into `Path#close()`:

```
Pack#close()
WindowCache.purge(Pack)
WindowCache#removeAll(Pack)
WindowCache#gc()
CleanupQueue#gc()
WindowCache#clear(PageRef<ByteWindow>)
WindowCache#close(Pack)
Pack#endWindowCache()
Pack#close()
```

---

Are you using memory mapped files, by any chance? (i.e. `WindowCacheConfig#setPackedGitMMAP(boolean)`). Memory mapped files remain open as long as the mapped buffers have not been reclaimed by GC. `System#gc()` may be of help here if that's the case.

---

JFYI, casting to `FileRepository` allows you to obtain the list of packfiles with far less reflection:

```java
Collection<Pack> packs = ((FileRepository) git.getRepository()).getObjectDatabase().getPacks();
```

with the caveat that `getPacks()` will attempt to locate packfiles on disk if not yet loaded into memory; it's not just dereferencing a field.

**Comment by arysin** [comment]
No, I don't use memory mapped files, at least I don't set any configurations for WindowCacheConfig - just using JGit pretty much out of the box.
Thanks for the tip on getting packs - it works and code is cleaner now.

**Comment by mfick-nvidia** [comment]
I think that the following connection is an asynchronous one:

`WindowCache#gc() -> CleanupQueue#gc()`

i.e. CleanupQueue#gc may or may not be called sometime in the future, but I don't think there are any guarantees as to when. Likely not in your situation, but in a multi threaded environment, in theory another thread might still be accessing the file and that in theory can also prevent the WindowCache from removing all entries for the Pack file, which would also prevent the gc from completing.

I think this might have been a typo:
```
Pack#endWindowCache()
Pack#close()
```
I believe it should actually be:
```
Pack#endWindowCache()
Pack#doClose()
```
Of course, `endWindowCache()` will only call `doClose()` under the "perfect" conditions, when all references to the pack are effectively cleared and there are no outstanding operations on it.

So, your expectation might be reasonable in your exact scenario, but I would say it would be the exception rather than the rule to have this expectation given the current code. What you are finding is that it might not actually even be enough of an exception, and that could be why it isn't working for you. 

`Pack.close()` is called on unix like OSes when the packfiles is already deleted, so perhaps it isn't really the right method for `PackDirectory.close()` to be calling? I suspect `PackDirectory.close()` would need to call something more aggressive to make it force close the file, like maybe calling `Pack.doClose()` eventually? The tricky part is that `Pack` actually jumps through a lot of hoops to prevent the fd from being closed until it is no longer being used, this helps ongoing transfers in other threads from potentially failing. Taking a step up, I also believe that `git.getRepository().close()` is not meant to be synchronous either, it might be operating on a repository from a cache which can also prevent it from closing synchronously.

**Comment by JoostK** [comment]
Thanks for the elaborate response, highly appreciated!

> I think that the following connection is an asynchronous one:
> 
> `WindowCache#gc() -> CleanupQueue#gc()`
> 
> i.e. CleanupQueue#gc may or may not be called sometime in the future, but I don't think there are any guarantees as to when. Likely not in your situation, but in a multi threaded environment, in theory another thread might still be accessing the file and that in theory can also prevent the WindowCache from removing all entries for the Pack file, which would also prevent the gc from completing.

This is quite sensitive indeed. If the `Pack` is held by the `WindowCache` then AFAICS it will go through `PageRef#kill()` to enqueue it into the `CleanupQueue`, which should then get cleaned up synchronously upon calling `CleanupQueue#gc()`. I do recognize that `Pack#endWindowCache()` may prevent this from happening if the pack file is still being accessed from other threads, but if that's not the case then closing is currently synchronous. Whether or not that is guaranteed or an implementation detail I can't tell.

---

> I think this might have been a typo:
> 
> ```
> Pack#endWindowCache()
> Pack#close()
> ```
> 
> I believe it should actually be:
> 
> ```
> Pack#endWindowCache()
> Pack#doClose()
> ```

Indeed it was, thanks for mentioning.

---

> Of course, `endWindowCache()` will only call `doClose()` under the "perfect" conditions, when all references to the pack are effectively cleared and there are no outstanding operations on it.
> 
> So, your expectation might be reasonable in your exact scenario, but I would say it would be the exception rather than the rule to have this expectation given the current code. What you are finding is that it might not actually even be enough of an exception, and that could be why it isn't working for you.

Note that I don't personally observe the "leak" as described by this issue's OP, but I'm taking meticulous care to close repository handles and actively purge them from `RepositoryCache` to drop their reference count to zero (it is being done in a write lock that is specific to the repository being closed, so other thread accesses are restricted _except_ for the thread that performs `RepositoryCache` purging). Without this effort, repositories will fail to be deleted on Windows platforms due to the dreaded `AccessDeniedException`.

> `Pack.close()` is called on unix like OSes when the packfiles is already deleted, so perhaps it isn't really the right method for `PackDirectory.close()` to be calling? I suspect `PackDirectory.close()` would need to call something more aggressive to make it force close the file, like maybe calling `Pack.doClose()` eventually? The tricky part is that `Pack` actually jumps through a lot of hoops to prevent the fd from being closed until it is no longer being used, this helps ongoing transfers in other threads from potentially failing.

This sounds reasonable and I totally understand the complexity here.

> Taking a step up, I also believe that `git.getRepository().close()` is not meant to be synchronous either, it might be operating on a repository from a cache which can also prevent it from closing synchronously.

Indeed, `RepositoryCache#close()` must be used to release its handle, if the `RepositoryCache` has had its hands on the `Repository`. I'm not sure if this may be why the OP is having issues, as the `Git` facade nor `CloneCommand` doesn't open repositories through `RepositoryCache#open` AFAICS.

**Comment by mfick-nvidia** [comment]
> If the Pack is held by the WindowCache then AFAICS it will go through PageRef#kill() to enqueue it into the CleanupQueue, which should then get cleaned up synchronously upon calling CleanupQueue#gc()

I think you can only count on this with hard references, if the system is using SoftReferences, then enqueuing the reference will only happen when the JVM garbage collector decides to do this, which may never happen.

**Comment by JoostK** [comment]
> > If the Pack is held by the WindowCache then AFAICS it will go through PageRef#kill() to enqueue it into the CleanupQueue, which should then get cleaned up synchronously upon calling CleanupQueue#gc()
> 
> I think you can only count on this with hard references, if the system is using SoftReferences, then enqueuing the reference will only happen when the JVM garbage collector decides to do this, which may never happen.

If it were only up to the garbage collector then yes, but `SoftRef#kill()` manually enqueues the `SoftReference` into the `ReferenceQueue` to make it eagerly visible to `SoftCleanupQueue#gc()`.
